### DISCUSSION TITLE
3D Graphing Issue - Senior Design Project

### DISCUSSION THREAD

**Discussion by chrisbergen4** [cooperrc/computational-mechanics#125] [General]
Hey, currently trying to graph an FFT in python for my senior design project. We are trying to see how the FFT chart progresses over time--thus our three dimensions are <time, frequency, amplitude> to graph. Cant seem to get any 3D stem plots to work with our arrays of data. Was wondering if anyone had any suggestions for how to plug the following code into a 3D chart:

X = np.abs(np.array([X1, X2, X3, X4,X5,X6,X7,X8,X9,X10,X11,X12,X13,X14,X15,X16,X17,X18]))
freq = np.array([freq1, freq2,freq3,freq4,freq5,freq6,freq7,freq8,freq9,freq10,freq11,freq12,freq13,freq14,freq15,freq16,freq17,freq18])
time = np.array([3000,4000,5000,6000,7000,8000,9000,10000,11000,12000,13000,14000,15000,16000,17000,18000,19000,20000])

<img width="577" alt="Screen Shot 2022-10-31 at 1 43 11 PM" src="https://user-images.githubusercontent.com/112510374/199073988-20f8bc90-0078-4a77-a3a6-3eaec7deec32.png">

**Comment by chrisbergen4** [comment]
above is an example of the type of 3D stem plot we want this data to work with

**Reply by chrisbergen4** [reply]
z = []
time = 2 #seconds
sr = 10000
N = time * sr
for i in range(N):
  t = i / sr
  x = np.sin(2*np.pi*4*(t**1.5))
  z.append(x)

n = np.arange(N)
time = n/sr
plt.plot(time, z)
plt.ylabel("Amplitude")
plt.xlabel("Time (Seconds)")
plt.title("Continuously Changing Frequency");


##normal FFT

from numpy.fft import fft, ifft

X = fft(z)

N = len(X)
n = np.arange(N)
T = N/sr
freq = n/T 
plt.tight_layout()
plt.figure(figsize = (4, 3))

plt.stem(freq, np.abs(X), 'b', \
         markerfmt=" ", basefmt="-b")
plt.xlabel('Freq (Hz)')
plt.ylabel('FFT Amplitude |X(freq)|')
plt.xlim(0, 10)
plt.title("Normal Fast Fourier Transformation")

plt.show()

##create 3D FFT

from numpy.fft import fft, ifft

X1 = fft(z[0:3000])
N = len(X1)
n = np.arange(N)
T = N/sr
freq1 = n/T 
X2 = fft(z[1000:4000])
N = len(X2)
n = np.arange(N)
T = N/sr
freq2 = n/T 
X3 = fft(z[2000:5000])
N = len(X3)
n = np.arange(N)
T = N/sr
freq3 = n/T 
X4 = fft(z[3000:6000])
N = len(X4)
n = np.arange(N)
T = N/sr
freq4 = n/T 
X5 = fft(z[4000:7000])
N = len(X5)
n = np.arange(N)
T = N/sr
freq5 = n/T 
X6 = fft(z[5000:8000])
N = len(X6)
n = np.arange(N)
T = N/sr
freq6 = n/T 
X7 = fft(z[6000:9000])
N = len(X7)
n = np.arange(N)
T = N/sr
freq7 = n/T 
X8 = fft(z[7000:10000])
N = len(X8)
n = np.arange(N)
T = N/sr
freq8 = n/T
X9 = fft(z[8000:11000])
N = len(X9)
n = np.arange(N)
T = N/sr
freq9 = n/T 
X10 = fft(z[9000:12000])
N = len(X10)
n = np.arange(N)
T = N/sr
freq10 = n/T 
X11 = fft(z[10000:13000])
N = len(X11)
n = np.arange(N)
T = N/sr
freq11 = n/T 
X12 = fft(z[11000:14000])
N = len(X12)
n = np.arange(N)
T = N/sr
freq12 = n/T 
X13 = fft(z[12000:15000])
N = len(X13)
n = np.arange(N)
T = N/sr
freq13 = n/T 
X14 = fft(z[13000:16000])
N = len(X14)
n = np.arange(N)
T = N/sr
freq14 = n/T 
X15 = fft(z[14000:17000])
N = len(X15)
n = np.arange(N)
T = N/sr
freq15 = n/T 
X16 = fft(z[15000:18000])
N = len(X16)
n = np.arange(N)
T = N/sr
freq16 = n/T 
X17 = fft(z[16000:19000])
N = len(X17)
n = np.arange(N)
T = N/sr
freq17 = n/T 
X18 = fft(z[17000:20000])
N = len(X18)
n = np.arange(N)
T = N/sr
freq18 = n/T 

plt.tight_layout()
plt.figure(figsize = (8, 6))

X = np.abs(np.array([X1, X2, X3, X4,X5,X6,X7,X8,X9,X10,X11,X12,X13,X14,X15,X16,X17,X18]))
freq = np.array([freq1, freq2,freq3,freq4,freq5,freq6,freq7,freq8,freq9,freq10,freq11,freq12,freq13,freq14,freq15,freq16,freq17,freq18])
time = np.array([3000,4000,5000,6000,7000,8000,9000,10000,11000,12000,13000,14000,15000,16000,17000,18000,19000,20000])

**Comment by cooperrc** [comment]
I think the plot you're looking for is called a [waterfall plot](https://en.wikipedia.org/wiki/Waterfall_plot)

e.g. 

![waterfall plot example for FFT](https://upload.wikimedia.org/wikipedia/commons/6/6e/Aaronia_Spectrum_Analyzer_Software.jpg)

**Reply by cooperrc** [reply]
This blog post looks like it could be a good start, 

[short time Fourier transform using Python+NumPy](https://kevinsprojects.wordpress.com/2014/12/13/short-time-fourier-transform-using-python-and-numpy/)
